### PR TITLE
Nebula nodejs SDK implementation

### PR DESCRIPTION
**Pull request by wujjpp**
# Nebula Nodejs SDK

This repository provides Nebula client API in Nodejs.

## Features

- Muti-Server Support
- Auto-reconnection support

  Client will try to reconnect forever, until the server is available again.  

- Connection pool support
- Disconnection detection

  A heartbeat mechanism is implemented, client will send ping to server each `pingInterval` ms for detect connective

- Thrift enhancement

  fix auto reconnect issue[#2407](https://github.com/apache/thrift/pull/2407)  

  fix performance issue in huge data scene[#2483](https://github.com/apache/thrift/pull/2483)

### Connection Options

| parameter      | type     | description                                                                     |
| -------------- | -------- | ------------------------------------------------------------------------------- |
| servers        | string[] | nebula servers                                                                  |
| userName       | string   | username for login                                                              |
| password       | string   | password for login                                                              |
| poolSize       | number   | Pool size for each server(Optional, default：5)                                 |
| bufferSize     | number   | Command cache in offline or before established connect (Optional, defaul: 2000) |
| executeTimeout | number   | Command executing timeout in ms (Optional, default：10000)                      |
| pingInterval   | number   | for keepalive, ping duration in ms, (Optional, default：60000）                 |

#### Install

For compiling C++ native module, `node-gyp` is required, you can install `node-gyp` by `npm install -g node-gyp`

```shell
npm install nebula-nodejs --save --unsafe-perm
```

#### Simple and convenient API

```typescript
// ESM
import { createClient } from 'nebula-nodejs'

// CommonJS
// const { createClient } = require('nebula-nodejs')

// Connection Options
const options = {
  servers: ['ip-1:port','ip-2:port'],
  userName: 'xxx',
  password: 'xxx',
  database: 'database name',
  poolSize: 5,
  bufferSize: 2000,
  executeTimeout: 15000,
  pingInterval: 60000
}

// Create client
const client = createClient(options)

// Execute command
// 1. return parsed data (recommend)
const response = await client.execute('GET SUBGRAPH 3 STEPS FROM -7897618527020261406')
// 2. return nebula original data
const responseOriginal = await client.execute('GET SUBGRAPH 3 STEPS FROM -7897618527020261406', true)

```

#### Events

| parameter         | description                                  |
| ----------------- | -------------------------------------------- |
| sender            | the individual connection in connection pool |
| error             | Nebula Error                                 |
| retryInfo         | Retry information                            |
| retryInfo.delay   | delay time                                   |
| retryInfo.attempt | total attempts                               |

```javascript
const client = createClient(options)

// connection is ready for executing command
client.on('ready', ({sender}) => {

})

// error occurs
client.on('error', ({ sender, error }) => {

})

// connected event
client.on('connected', ({ sender }) => {

})

// authorized successfully
client.on('authorized', ({ sender }) => {

})

// reconnecting
client.on('reconnecting', ({ sender, retryInfo }) => {

})

// closed
client.on('close', { sender }) => {

}
```

### About hash64 function

`nebula-nodejs` exports `hash64` function for converting `string` to `string[]`, it's based on `MurmurHash3`.

```javascript
// ESM
import { hash64 } from 'nebula-nodejs'

// CommonJS
// const { hash64 } = require('nebula-nodejs')

const results = hash64('f10011b64aa4e7503cd45a7fdc24387b')

console.log(results)

// Output:
// ['2852836996923339651', '-6853534673140605817']
```

### About Int64

nodejs cannot repreent `Int64`, so we convert `Int64` bytes to `string`

```javascript
// ESM
import { bytesToLongLongString } from 'nebula-nodejs'

// CommonJS
// const { bytesToLongLongString } = require('nebula-nodejs')

const s = '-7897618527020261406'

const buffer = [146, 102, 5, 203, 5, 105, 223, 226]
const result = bytesToLongLongString(buffer)

// result equals s
```

### Build

```shell
git clone https://github.com/vesoft-inc/nebula-node.git
cd nebula-node
npm install --unsafe-perm
npm run build
```

### Unit Test

```shell
npm run build
npm run test
```

### Unit Test Coverage

```shell
npm run coverage
```

### Publish

```shell
npm run build
cd dist
npm publish
```

## TODO

Not implemented data type for auto parser

| Data Type | property name in nebula response |
| --------- | -------------------------------- |
| NMap      | mVal                             |
| NSet      | uVal                             |
| DataSet   | gVal                             |

### CODE REVIEW

**Comment by Reid00** [comment]
@wey-gu here is the pr, pls take a look.

**Comment by wey-gu** [comment]
> @wey-gu here is the pr, pls take a look.

Thank you @Reid00 @wujjpp! Great Job! You even fixed two issues on upstream apache thrift 👍🏻. We will look into it.

Best Regards, Wey

File: README.md [hunk 1]
```diff
@@ -1,25 +1,189 @@
...
+// Connection Options
+const options: ClientOption = {
+  servers: ['ip-1:port','ip-2:port'],
+  userName: 'xxx',
+  password: 'xxx',
+  database: 'database name',
+  poolSize: 5,
+  bufferSize: 2000,
+  executeTimeout: 15000,
+  pingInterval： 60000
```
Comment (wey-gu) [inline README.md]: Kindly help fixed the colon here ;)

Comment (wujjpp) [reply README.md]: @wey-gu Fixed

File: package.json [hunk 2]
```diff
@@ -1,29 +1,113 @@
...
+    "mocha": "^8.4.0",
+    "moment": "^2.29.1",
+    "nyc": "^15.1.0",
+    "open": "^7.4.2",
+    "rimraf": "^3.0.2",
+    "sinon": "^9.2.4",
+    "sinon-chai": "^3.7.0",
+    "source-map-support": "^0.5.19",
+    "typescript": "4.1.5",
+    "uuid": "^8.3.2",
```
Comment (jerryliang64) [inline package.json]: should these lib be in the package.json  `dependencies` instead of `devDependencies` ?

File: package.json [hunk 3]
```diff
@@ -1,29 +1,113 @@
...
+    "eslint-plugin-jsdoc": "^32.3.4",
+    "eslint-plugin-node": "^11.0.0",
+    "eslint-plugin-prefer-arrow": "^1.2.2",
+    "eslint-plugin-promise": "^4.2.1",
+    "eslint-plugin-standard": "^5.0.0",
+    "esm": "^3.2.25",
+    "inquirer": "^8.1.1",
+    "mkdirp": "^0.5.1",
+    "mocha": "^8.4.0",
+    "moment": "^2.29.1",
```
Comment (jerryliang64) [inline package.json]: Just for advice: Is it better to use day.js instead of monent.js  for reducing the package size? https://medium.datadriveninvestor.com/https-medium-com-sabesan96-why-you-should-choose-day-js-instead-of-moment-js-9cf7bb274bbd

File: package.json (hunk 2 above)
Comment (wujjpp) [reply package.json]: 1. `​browser-or-node`, `isomorphic-ws`, `node-int64`, `q`, `ws` dependent by thrift, because we have copy & modified thrift in local package, so we should add thoese libs as `dependencies `
2. `bindings`, `lodash`  dependent by `nebula-node`
3. all other libs treat as `devDependencies`

File: package.json (hunk 3 above)
Comment (wujjpp) [reply package.json]: `moment` is useless, I will remove it

File: package.json (hunk 2 above)
Comment (jerryliang64) [reply package.json]: 👌，I just find all the libs in the devDependencies.

File: src/nebula/Client.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+import NebulaError from './NebulaError'
+import { EventEmitter } from 'events'
+
+let AsyncResource
+let executionAsyncId
+let isSupported = false
+
+try {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const asyncHooks = require('async_hooks')
```
Comment (jerryliang64) [inline src/nebula/Client.ts:17]: I'm curious about why not just import `import async_hooks` directly, is it possible that the module can't be imported?

Comment (jerryliang64) [reply src/nebula/Client.ts:17]: BTW, I find you use `import` other place:  https://github.com/vesoft-inc/nebula-node/blob/5d71bfd743ae512d2dc490b47897022cce5d6d4b/src/nebula/types.ts

Comment (wujjpp) [reply src/nebula/Client.ts:17]: `async_hooks` module added in nodejs v8.1.0 and in fact `nebula-node` can be run on `nodejs` below ` v8.1.0`, although I have specifed v10.0.0 in `package.json`. we use `require` for detecting the target node compatibility. If it not support, our module can also working fine.

Comment (wujjpp) [reply src/nebula/Client.ts:17]: Directory Layout:
.
├── /**_dist_**/                                    # Build results for publishing (this folder generated after `npm run build`, its for publish)
├── /specs/                                 # Misc knowledge and docs
├── /src/                                     # Source Dir 
├── /tests/                                  # Unit test case, if you have nebula server, Put your test case in this folder for testing
├── /tools/                                 # Tiny build system
└── ....                                         # Supporting files

Build process:
image

Build & Publish:
In short:  `typescript` -> `ESM` -> `CommonJS` -> `Copy & Generate necessary assets` -> `publish`

**Comment by wujjpp** [comment]
@wey-gu 这个PR还合吗？

**Comment by jerryliang64** [comment]
> 

We'll deal with it in this week, thanks for your following🤝

**Comment by wey-gu** [comment]
> @wey-gu 这个PR还合吗？

Sorry for the late response, sure we will, hopefully it could be done in short time as aligned with @nianiaJR

**Review by jerryliang64 — COMMENTED** [review]
@wujjpp Is there any plan to make up the unit test/example in future?
image
image

I find it only like a `todo` demo about test?

File: src/nebula/types.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+interface ClientOption {
+  // nebula服务器列表
+  servers: string[] | Endpoint[];
+  // 用户名
+  userName: string;
+  // 密码
+  password: string;
+  // 数据库名称
+  database: string;
```
Comment (jerryliang64) [inline src/nebula/types.ts]: It's for space? maybe it's more clear to rename it to space ?

**Review by jerryliang64 — COMMENTED** [review]
I've simply tried this lib, work fine. Because of it's too big changes, it's important to make up the enough unit test/examples for long-term maintaining. Thanks for your contributions. I'll approved it and make it merged soon🤝

File: src/nebula/types.ts (hunk 5 above)
Comment (wujjpp) [reply src/nebula/types.ts]: @nianiaJR rename `database`  to `space` done

**Comment by wujjpp** [comment]
> @wujjpp Is there any plan to make up the unit test/example in future? image image
> 
> I find it only like a `todo` demo about test?

 I have a full test case located at `tests/nebula.test-disabled`, but I didn't have `nebula` enviroment for testing in github env.
So, I rename `nebula.test.ts` to `nebula.test-disabled`, you can revert its name and add unit tests for testing.
BTW, those test cases are based on our nebula production server.

```ts
/* eslint-disable max-len */
/**
 * Created by Wu Jian Ping on - 2021/06/09.
 */

import { expect } from 'chai'
import createClient, { ClientOption } from '../dist'

const relationServer: ClientOption = {
  servers: ['10.0.1.100:9669', '10.0.1.101:9669', '10.0.1.102:9669'],
  userName: 'xxxxx',
  space: 'xxxxx',
  database: 'partion1',
  pingInterval: 5000,
  poolSize: 2
}

const riskInfoServer: ClientOption = {
  servers: ['10.0.1.100:9669', '10.0.1.101:9669', '10.0.1.102:9669'],
  userName: 'xxxxx',
  space: 'xxxxx',
  database: 'partion2',
  pingInterval: 5000,
  poolSize: 2
}

const commands = {
  cmd1: 'GET SUBGRAPH 2 STEPS FROM -7897618527020261406',
  cmd2: 'fetch prop on Company -7897618527020261406',
  cmd3: 'go from 815677140545765099 over Relation yield Relation._src as src, Relation._dst as dst, Relation.relation_name as relation_name, Relation.prop1 as prop1, Relation.prop2 as prop2, Relation.prop3 as prop3, Relation.prop4 as prop4, $^.Person.name as src_p_name, $^.Company.name as src_c_name,  $$.Person.name as dst_p_name, $$.Company.name as dst_c_name | limit 1000',
  cmd4: 'find noloop path from -4075961126534726064 to 815677140545765099 over Relation bidirect upto 2 steps',
  cmd5: 'fetch prop on Relation -4075961126534726064->815677140545765099@-5101473608195470769'
}

describe('nebula', () => {
  // eslint-disable-next-line prefer-arrow/prefer-arrow-functions
  it('test-case-1', async () => {
    const client = createClient(relationServer)

    const response1 = await client.execute(commands.cmd1)

    expect(response1.data?._vertices?.length).greaterThan(0)
    expect(response1.data?._edges?.length).greaterThan(0)

    await client.close()
  })

  it('test-case-2', async () => {
    const client = createClient(relationServer)

    const response1 = await client.execute(commands.cmd2)

    expect(response1.data?.vertices_?.length).greaterThan(0)

    await client.close()
  })

  it('test-case-3', async () => {
    const client = createClient(riskInfoServer)

    const response1 = await client.execute(commands.cmd3)
    expect(response1.data?.src?.length).greaterThan(0)

    await client.close()
  })

  it('test-case-4', async () => {
    const client = createClient(riskInfoServer)

    const response1 = await client.execute(commands.cmd4)

    expect(response1.data?.path?.length).greaterThan(0)

    await client.close()
  })

  it('test-case-5', async () => {
    const client = createClient(riskInfoServer)

    const response1 = await client.execute(commands.cmd5)

    expect(response1.data?.edges_?.length).greaterThan(0)

    await client.close()
  })

  it('test-promise-all', async () => {
    const client = createClient(relationServer)
    const [resp1, resp2, resp3] = await Promise.all([
      client.execute('GET SUBGRAPH 1 STEPS FROM -7897618527020261406', false),
      client.execute('GET SUBGRAPH 2 STEPS FROM -7897618527020261406', false),
      client.execute('GET SUBGRAPH 3 STEPS FROM -7897618527020261406', false)
    ])

    console.log(`resp11:${resp1.data._vertices.length}, ${resp1.data._edges.length}`) // eslint-disable-line
    console.log(`resp21:${resp2.data._vertices.length}, ${resp2.data._edges.length}`) // eslint-disable-line
    console.log(`resp31:${resp3.data._vertices.length}, ${resp3.data._edges.length}`) // eslint-disable-line

    await client.close()
  })

  after(async () => {
    process.exit()
  })
})
```

**Review by jerryliang64 — APPROVED** [review]
LGTM
